docs/architecture/architecture.md

```diff
@@ -0,0 +1,378 @@
+
+`qemu-lite` improvements comes through a new `pc-lite` machine type, mostly by:
+- Removing many of the legacy hardware devices support so that the guest kernel does not waste
+time initializing devices of no use for containers.
+- Skipping the guest BIOS/firmware and jumping straight to the Clear Containers kernel.
+
+#### Agent
+
+['agent'](https://github.com/clearcontainers/agent) is a daemon running on the guest as an
+agent for managing containers and processes potentially running within those containers.
+
+The `agent` execution unit is the pod. An `agent` pod is a container sandbox defined
+by a set of namespaces (UTS, PID, mount and IPC). Although a pod can hold several containers,
+`cc-runtime` always runs a single container per pod. **fixme** incorrect<--
+
+**todo** add details on the agent protocol
+
+#### Runtime
+
+`cc-runtime` is an OCI compatible container runtime and is responsible for handling all
+commands specified by [the OCI runtime specification](https://github.com/opencontainers/runtime-spec)
+and launching `cc-shim` instances.
+
+Here we will describe how `cc-runtime` handles the most important OCI commands.
+
+##### `create`
+
+When handling the OCI `create` command, `cc-runtime` goes through the following steps:
+
+1. Create the container namespaces (Only the network and mount namespaces are currently supported),
+according to the container OCI configuration file.
+2. Spawn the `cc-shim` process and have it wait on a couple of temporary pipes for:
+   * A `cc-proxy` created file descriptor (one end of a socketpair) for the shim to connect to.
+   * The container `agent` sequence numbers for at most 2 I/O streams (One for `stdout` and `stdin`, one for `stderr`).
+   The `agent` uses those sequence numbers to multiplex all streams for all processes through one serial interface (The
+   virtio I/O serial one).
+3. Run all the [OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hooks) in the container namespaces,
+as described by the OCI container configuration file.
+4. **fixme** [Set up the container networking](https://github.com/clearcontainers/runtime/blob/master/documentation/architecture.md#networking).
+This must happen after all hooks are done as one of them is potentially setting
+the container networking namespace up.
+5. Create the virtual machine running the container process. The VM `systemd` instance will spawn the `agent` daemon.
+6. Wait for the `agent` to signal that it is ready.
+7. Send the pod creation command to the `agent`. The `agent` pod is the container process sandbox.
+8. Send the allocateIO command to the proxy, for getting the `agent` I/O sequence numbers described in step 2.
+9. Pass the `cc-proxy` socketpair file descriptor, and the I/O sequence numbers to the listening cc-shim process through the dedicated pipes.
+10. The `cc-shim` instance is put into a stopped state to prevent it from doing I/O before the container is started.
```

Comment:
8 and 9 can be removed with 3.0
10 is not accurate. The shim is alive after a create.